gui: kb shortcuts: fix incorrect wx key codes for "plain" keys

avih opened the pull request:
This fixes two issues which were observed on windows:
- plain letters: wx passes the capital code (e.g. 81 instead of 113 for 'q')
- symbol-only keys: just incorrect code (e.g. 219 instead of 91 for '[')

For both these cases, holding e.g. shift correctly has the shift flag set, and
the key code remains the same.

This should allow using letters and symbols at PCSX2_keys.ini

Note: this should have supposedly been fixed at r4918 ( 851bfba ), but neither
that build from the buildbot archive, nor the official 1.2.1 release (r5875),
nor the current builds have this working, which is really weird (tested on win8)

avih commented:
@turtleli IIRC you touched some wx key codes stuff. What do you think of this PR?

- Can you reproduce the issue that it's impossible to use plain chars (letters or symbols which are not special wx key codes) at `PCSX2_keys.ini`?
- If yes, does this PR fix it for you?
- Regardless, is this PR reasonable? or do you have a better/more-correct approach here?

avih commented:
Just noticed that the symbols "errors" are actually just having the MSb (128) set. I.e. subtracting 128 ends up with the correct code. So maybe we should just check if the keycode is above .. 32+128, and subtract 128?

Also, I now notice that with the patch, `shift-q` doesn't seem to work, but it was working for me before the patch. Both before and after the patch the console printed e.g.: `assigning Shift+Q (instead of Shift+F2)`. So the assignment seems incorrect, but because the wx key code was also incorrectly for upper case, it ended up working.

So this patch so far fixes symbol and plain letters without shift (need to test other modifiers too), but breaks shift+plain-letters.

avih commented:
OK, I think I got everything working. AFAICT there were actually 3 different issues (see the commit message for details). This is ugly stuff.

@turtleli despite the time I put into "fixing" it, I'd take a better fix without all those hacks over my code without hesitation... if you can think of something better... let me know how.

turtleli commented:
Seems like a Windows-only issue.
Linux with onepad - cannot reproduce.
Windows with Pokopom or Lilypad - can reproduce.

I think the problem is:
Input - Virtual keycodes (https://msdn.microsoft.com/en-us/library/windows/desktop/dd375731(v=vs.85).aspx)
Shortcuts - not virtual keycodes, I don't know what form it's in, haven't looked at it (yet).

So the shortcuts simply aren't able to match.

avih commented:
> Windows with ... Lilypad - can reproduce.

What do you mean "can reproduce"? for which case? I have LilyPad set to use plain keys and it works correctly, and always had.

As far as I can tell, the issue is only with the key presses which PCSX2 tries to interpret for its own commands. I didn't see any indication that it affects LilyPad (though I haven't looked too deeply).

turtleli commented:
I mean I can reproduce the issue of plain character keyboard shortcuts not working on Windows. I checked Lilypad to see what it was doing, and it was simply passing along the virtual key codes it received if it wasn't using the keys itself.

avih commented:
I'm not sure I understand what you imply.
- Do you think the patch breaks something with LilyPad on Windows, or anything on linux, or anything else?
- Regardless if it break anything, do you think there is/could be a better approach than this patch takes?

turtleli commented:
My comments were about "before patch" state - Linux works, Windows doesn't.

As for the patch - I don't think it'll break anything on Windows (seems to work), but it might break Linux, I haven't checked.

I'll think if there's a better approach after some sleep.

avih commented:
> I'll think if there's a better approach after some sleep.

Same :)

avih commented:
Like you said, sleep helped :)

So I think I get it. I didn't look at your link earlier, but now I have, and I now see that my translation is just translating the rest of the VK codes which TranslateVKToWXK doesn't handle. So I think the solution should be to enhance TranslateVKToWXK.

There would remain an issue that only 4 of the symbols-only keys are consistent between regions/keyboards (such as `VK_OEM_PERIOD` for `.`) but the other 7 aren't (`VK_OEM_1` for `;` to `VK_OEM_7` for `'`).

I think we can live with translating always the US layout, and also, users with non US KB which want to use those keys would hopefully be able to enable the dev console (even in release), press the key, and PCSX2 will print the char it expects to see at `PCSX2_keys.ini`.

Or, maybe there's a wx function which can translate VK codes to char codes, which we could use inside/instead of TranslateVKToWXK.

turtleli commented:
You need more sleep, WIndows 2000 was released before WIndows XP :)

But I think we can try`MapVirtualKey` first and see how well that works.

avih commented:
I moved the extra translations into `WinKeyCodes.cpp` where it's much more appropriate, and added a print to the console if DevCon is enabled, so whatever this prints, the user should be able to copy-paste it into PCSX2_keys.ini and (fingers crossed) it should work.

I think this should be OK to push now. I'll change it later it into a translation table to make it quicker.

@turtleli you OK with this?

turtleli commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -294,6 +294,20 @@ void Pcsx2App::PadKeyDispatch( const keyEvent& ev )
+			(m_kevt.m_altDown ? L"Alt-" : L""),
+			(m_kevt.m_controlDown ? L"Ctrl-" : L""),
+			(m_kevt.m_shiftDown ? L"Shift-" : L""),
+			strFromCode.c_str()
```
Needs to be WX_STR(strFromCode), otherwise it outputs garbage on Windows, and fails to compile on Linux. That's probably why the extra %ld didn't work properly.

turtleli commented:
I'm fine with this (after it's fixed). Briefly Windows and Linux tested.

I tested the copy-paste into the ini on a UK keyboard and that works. Would be better if the characters matched, but people don't change their shortcuts everyday... I think.

avih commented:
> I tested the copy-paste into the ini on a UK keyboard and that works.

That's nice. Thanks.

> Would be better if the characters matched, but people don't change their shortcuts everyday... I think.

Agreed, but it apparently works also for non US KBs, and it's been way too much hassle already, because of an unintended removal of this function at wxWidgets - http://trac.wxwidgets.org/ticket/12508 (apparently one dev change the semantics so he hid it till a compatible function is provided, but then wx3 was released with the function hidden, and then they didn't see the need to return it since a version was already released without it... so meh).

Anyway, the WX_STR indeed fixed it, thanks, and I also used the accelerator print function after setting the CTOR flags correctly instead of rolling out our own. That's better.

I'll squash and commit. Thanks.

avih commented:
> I also used the accelerator print function after setting the CTOR flags correctly instead of rolling out our own. That's better.

This also returns e.g. `Alt+Shift+a` while the own print had `Alt-Shift-a`, but both work at the ini so that's fine.

turtleli commented:
Retested on Linux, seems fine.

avih commented:
I just tested tried the official 1.2.1 again (r5875) and also tried r4918 (which originally fixed plain letters and symbols), and on both of them symbols (e.g. `]`) with or without modifiers work, and also plain letters with modifiers work (like `Shift-q`), but plain letters without modifiers don't work. So basically the only thing which doesn't work in 1.2.1 is plain letters without modifiers.

The regression for the cases which did work in 1.2.1 (symbols, letters with modifiers) is from the move to wxWidgets 3.0. These cases now work again after this PR got merged.
